domains/certificates/Query.API/API/DataSyncSyncer/DataSyncService.cs

```diff
@@ -0,0 +1,123 @@
+    {
+        this.httpClient = httpClient;
+        this.logger = logger;
+    }
+
+    public void SetState(Dictionary<string, DateTimeOffset> state)
+    {
+        periodStartTimeDictionary = state;
+    }
+
+    public async Task<List<DataSyncDto>> FetchMeasurements(string GSRN, string meteringPointOwner,
+        CancellationToken cancellationToken)
+    {
+        var dateFrom = periodStartTimeDictionary[GSRN].ToUnixTimeSeconds();
+
+        var now = DateTimeOffset.UtcNow;
+        var midnight = new DateTimeOffset(now.Year, now.Month, now.Day, 0, 0, 0, TimeSpan.Zero).ToUnixTimeSeconds();
+        var result = new List<DataSyncDto>();
+
+        if (dateFrom < midnight)
+        {
+            try
+            {
+                result = await RequestDatasync(
+                    GSRN,
+                    new Period(
+                        DateFrom: dateFrom,
+                        DateTo: midnight
+                    ),
+                    meteringPointOwner,
+                    cancellationToken
+                );
+            }
+            catch (Exception e)
+            {
+                logger.LogInformation("An error occured: {error}, no measurements was fetched", e.Message);
+            }
+
+        }
+
+        SetNextPeriodStartTime(result, GSRN);
+        return result;
+    }
+
+    private void SetNextPeriodStartTime(List<DataSyncDto> measurements, string GSRN)
+    {
+        if (measurements.IsEmpty())
+        {
+            return;
+        }
+
+        var newestMeasurement = measurements.Max(m => m.DateTo);
+        periodStartTimeDictionary![GSRN] = DateTimeOffset.FromUnixTimeSeconds(newestMeasurement);
+    }
+
+    private async Task<List<DataSyncDto>> RequestDatasync(string gsrn, Period period, string meteringPointOwner,
+        CancellationToken cancellationToken)
+    {
+        logger.LogInformation(
+            "Fetching data in period from {from} to: {to}", period.DateFrom.ToString("MM/dd/yy H:mm:ss"), period.DateTo.ToString("MM/dd/yy H:mm:ss")
+        );
+
+        var url = $"measurements?gsrn={gsrn}&dateFrom={period.DateFrom}&dateTo={period.DateTo}";
+        httpClient.DefaultRequestHeaders.Authorization =
+            new AuthenticationHeaderValue("Bearer", GenerateToken(meteringPointOwner));
+
+        var response = await httpClient.GetAsync(url, cancellationToken);
+
+        return await response.Content.ReadFromJsonAsync<List<DataSyncDto>>(cancellationToken: cancellationToken)
+               ?? throw new Exception($"Fetch of measurements failed, base: {httpClient.BaseAddress} url: {url}");
+    }
+
+    private static string GenerateToken(string meteringPointOwner)
+    {
+        var claims = new Claim[]
+        {
+            new(JwtRegisteredClaimNames.UniqueName, "username"),
+            new(JwtRegisteredClaimNames.NameId, Guid.NewGuid().ToString()),
+            new("subject", meteringPointOwner),
+            new("actor", "actor"),
+            new("scope", "scope"),
```

Comment:
These do not have a valid value. We could change it so, "actor" is the same value as "subject" and then have the scopes "meteringpoints.read" and"measurements.read". To add multiple scopes I believe you have to make multiple Claim with the same key ("scope")